db/memtable.h

```diff
@@ -601,9 +607,21 @@ class MemTable {
                     MergeContext* merge_context, SequenceNumber* seq,
                     bool* found_final_value, bool* merge_in_progress);
 
-  // Always returns non-null and assumes certain pre-checks are done
+  // Always returns non-null and assumes certain pre-checks (e.g.,
+  // is_range_del_table_empty_) are done. This is only valid during the lifetime
+  // of the underlying memtable.
   FragmentedRangeTombstoneIterator* NewRangeTombstoneIteratorInternal(
       const ReadOptions& read_options, SequenceNumber read_seq);
+
+  // The fragmented range tombstones of this memtable.
+  // This is constructed when this memtable becomes immutable
+  // if !is_range_del_table_empty_.
+  std::unique_ptr<FragmentedRangeTombstoneList>
+      fragmented_range_tombstone_list_;
+
+  // Denotes whether a fragmented range tombstone list is cached with the
+  // memtable. This should only be set be immutable memtables.
+  std::atomic<bool> has_range_tombstone_list_;
```

Comment:
Do you see any possibility for an API (like `PrepareImmutable()` or something) that is called by `SwitchMemtable()` with lock released? I've also been thinking about madvise() with MADV_DONTNEED the unused memory at the same point.

Reply:
Also regarding this particular variable, I wonder if `MemTableListVersion::AddRangeTombstoneIterators()` should pass the info about immutability down to the `MemTable::NewRangeTombstoneIterator()`.

Reply:
> PrepareImmutable()
I think an API like this is great once we do multiple things in the prepare phase (like when we add `madvise()` calls). One question about `unused memory`, what are the don't-needed memory in the immutable memtable? Will they be accessed by threads holding old SuperVersion that still treat this memtable as a live memtable?

Reply:
> Also regarding this particular variable, I wonder if `MemTableListVersion::AddRangeTombstoneIterators()` should 
pass the info about immutability down to the `MemTable::NewRangeTombstoneIterator()`.

Ah yes, good point. I think a similar change in Get/MultiGet calls are also needed and doable. Then we can get rid of this atomic flag.